### PR TITLE
Add 3 vim operations

### PR DESCRIPTION
**Pull request by xianzhon**
As a vim user, in my daily work, I rely on below vim operations frequently.
Although in vnote, we have shortkeys ([[,]],[{,]}) to navigate by markdown title, below operations still be useful.  So I added it.

S (shift+s): delete current line and go to insert mode
{: go to previous paragraph
}: go to next paragraph

### CODE REVIEW

File: src/utils/vvim.cpp [hunk 1]
```diff
@@ -1029,6 +1029,15 @@ bool VVim::handleKeyPressEvent(int key, int modifiers, int *p_autoIndentPos)
             }
 
             setMode(VimMode::Insert);
+        } else if (modifiers == Qt::ShiftModifier) {
```
Comment (tamlok) [inline src/utils/vvim.cpp]: As far as I know, `S` is the same as `cc`, right? I have implemented the `cc`. I think you could just copy that logic here. By the way, we need to care about the `[count]` repeat number and the auto-indent, which has been handled in the `cc`.

File: src/utils/vvim.cpp [hunk 2]
```diff
@@ -2047,12 +2056,26 @@ bool VVim::handleKeyPressEvent(int key, int modifiers, int *p_autoIndentPos)
...
                 processCommand(m_tokens);
                 break;
-            } else if (hasActionToken() || !checkMode(VimMode::Normal)) {
-                // Invalid sequence.
-                break;
-            } else if (checkPendingKey(Key(Qt::Key_BracketLeft))) {
+            } else if (hasActionToken() && checkPendingKey(Key(Qt::Key_BracketLeft))) {
                 // [{, goto previous title at one higher level.
                 processTitleJump(m_tokens, false, -1);
+            } else if (checkMode(VimMode::Normal) || checkMode(VimMode::Visual) || checkMode(VimMode::VisualLine)) {
```
Comment (tamlok) [inline src/utils/vvim.cpp]: The Vim mode implementation is a little complicated in VNote, since we need to take of the `[count]` and other stuff. In VNote, we abstract the Vim operations like `[repeat][action][movement]`. For example, `2{`, `d2{`, `c{`, or `y{` is not considered and handled here. You could reference other movement's implementations for the details.

Comment (xianzhon) [reply src/utils/vvim.cpp]: Got it. Thanks for your suggestion, it's helpful!

**Comment by xianzhon** [comment]
@tamlok I already modified the implementation by considering the vim motion & movement.
And I did below test, same as vim. Please check it~
```
* test operation `S`
2S, 1000S

* test movement {,} 

d2{, d2}, 2d{, 2d}
2y{, 2y}, 2{, 2}, 100{
```

File: src/utils/vvim.cpp [hunk 3]
```diff
@@ -1029,6 +1101,12 @@ bool VVim::handleKeyPressEvent(int key, int modifiers, int *p_autoIndentPos)
             }
 
             setMode(VimMode::Insert);
+        } else if (modifiers == Qt::ShiftModifier) {
+            // S, delete line and go to insert mode
+            tryGetRepeatToken(m_keys, m_tokens);
+            addActionToken(Action::Change);
```
Comment (tamlok) [inline src/utils/vvim.cpp]: It is legal only when currently there is no Action token. Key press such as `dS` will crash here.

File: src/utils/vvim.cpp [hunk 4]
```diff
@@ -2047,12 +2125,14 @@ bool VVim::handleKeyPressEvent(int key, int modifiers, int *p_autoIndentPos)
                 addRangeToken(range);
                 processCommand(m_tokens);
                 break;
-            } else if (hasActionToken() || !checkMode(VimMode::Normal)) {
-                // Invalid sequence.
-                break;
-            } else if (checkPendingKey(Key(Qt::Key_BracketLeft))) {
+            } else if (hasActionToken() && checkPendingKey(Key(Qt::Key_BracketLeft))) {
```
Comment (tamlok) [inline src/utils/vvim.cpp]: `hasActionToken()`? `[{` now does not go to the upper level title. It just trigger the wrong action `ParagraphUp` instead. :(

Comment (xianzhon) [reply src/utils/vvim.cpp]: I see, I think it's hard to follow the vim spec and in same time add our own spec. So decided to close this PR, and make it as my own customization. Thanks for your help.

Comment (tamlok) [reply src/utils/vvim.cpp]: @xianzhon :) I have implemented these 3 operations according to yours (9ef59990dc6eee79d0b219bb625d51ab4df055ce). Thanks very much!

Comment (xianzhon) [reply src/utils/vvim.cpp]: @tamlok Nice! I read your commmit, your implementation is more like the vim style (I'm now reading the vim source code) and must be more efficient too! Good!!
